core/src/main/scala/caliban/relay/PaginationArgs.scala

```diff
@@ -0,0 +1,91 @@
+package caliban.relay
+
+import caliban.CalibanError
+import zio._
+
+object Pagination {
+  import PaginationCount._
+  import PaginationCursor._
+
+  def apply[C: Cursor](
+    args: PaginationArgs[C]
+  ): ZIO[Any, CalibanError, Pagination[C]] =
+    apply(args.first, args.last, args.before, args.after)
+
+  def apply[C: Cursor](
+    first: Option[Int],
+    last: Option[Int],
+    before: Option[String],
+    after: Option[String]
+  ): ZIO[Any, CalibanError, Pagination[C]] =
+    ZIO
+      .mapParN(
+        validateFirstLast(first, last),
+        validateCursors(before, after)
+      )((count, cursor) => new Pagination[C](count, cursor))
+      .parallelErrors
+      .mapError((errors: ::[String]) => CalibanError.ValidationError(msg = errors.mkString(", "), explanatoryText = ""))
+
+  private def validateCursors[C: Cursor](
+    before: Option[String],
+    after: Option[String]
+  ) =
+    (before, after) match {
+      case (Some(_), Some(_)) =>
+        ZIO.fail("both before and after may not be set")
+      case (Some(x), _)       =>
+        ZIO.fromEither(Cursor[C].decode(x)).map(Before(_))
+      case (_, Some(x))       =>
+        ZIO.fromEither(Cursor[C].decode(x)).map(After(_))
+      case (None, None)       => ZIO.succeed(NoCursor)
+    }
+
+  private def validateFirstLast(first: Option[Int], last: Option[Int]) =
+    (first, last) match {
+      case (None, None)       =>
+        ZIO.fail("first and last cannot both be empty")
+      case (Some(_), Some(_)) =>
+        ZIO.fail("both first and last cannot be set")
```

Comment:
nit: "first and last cannot both be set"

Reply:
👍 2c1a2cfd